# IoTアプリケーションの開発効率の向上を目的とした開発者によるコードの変更を動的に適用する方式の提案と実装

このノートブックは、情報処理学会の[第207回SE研究発表会](http://www.ipsj.or.jp/kenkyukai/event/se207.html)において口頭発表予定の研究報告論文「IoTアプリケーションの開発効率の向上を目的とした開発者によるコードの変更を動的に適用する方式の提案と実装」のための実験について詳細を記録・公開するためのリポジトリである。

計測方法について、共著者の北九州市立大学の山崎進氏による記事「[mix_tasks_upload_hotswap の Hexライブラリ版を試す](https://qiita.com/zacky1972/items/f0b47eded7c902008871)」、および、共著のプロセスを通じてご指導いただきました。感謝いたします。

## 更新対象とするIoTアプリケーション

各方式の比較検討に用いるアプリケーションは、本リポジトリの[target_app](./target_app)ディレクトリに収められている、IoTデバイス上で動作するEchoサーバが更新対象のアプリケーションである。

## 比較検討する方式

上記のIoTアプリケーションコードの変更をIoTデバイスに適用する方式として、以下の3つについて比較検討する。

1. ファームウェアイメージの全体を適用する方式
2. ファームウェアイメージの差分を適用する方式
3. アプリケーションコードを動的に適用する方式（提案方式）

それぞれの詳細については、研究報告論文を参照されたい。

## 比較検討の方法

各方式を比較するための指標として、開発者によるコードの変更をデバイスに適用し、更新対象のIoTアプリケーションがTCP通信可能となるまでの時間を用いる。

方式を実装するためのIoTデバイス開発プラットフォームとして、[Nerves](https://www.nerves-project.org/)を用いる。筆者らによる提案方式3.の実装を加えると、上記の方式をすべて実装しているのがNervesだからである。ハードウェアとしては、Raspberry Piを用いる。

方式1.および2.については、スクリプト[measure.sh](./measure.sh)を用いて、以下の時間を計測する。

1. ファームウェアイメージの生成
2. ファームウェアイメージのデバイスへのアップロード
3. デバイスの再起動

方式3.については、提案方式を実装した`mix upload.hotswap`コマンドの実行時間を、`time(1)`によって計測する。

また、それぞれの方式について、アップデート対象となるファームウェアあるいはコードの容量についても付記する。

## 実験環境

### 開発環境

* Model: MacBook Pro 13-inch, 2018
* CPU: 2.7 GHz Quad-Core Intel Core i7
* Memory: 16 GB 2133 MHz LPDDR3

### IoTデバイス

* Raspberry Pi 3 Model B
* 開発環境との通信: 有線LAN
* プラットフォーム: Nerves 1.7.2

このNotebookでは、Nervesの開発環境の整備については解説しない。「[ElixirでIoT#4.1：Nerves開発環境の準備（2020年11月版）](https://qiita.com/takasehideki/items/88dda57758051d45fcf9)」等の記事を参照されたい。

## 実験

実験では、筆者らが用意したスクリプト[measure.sh](./measure.sh)を用いる。

### 1. ファームウェアイメージ全体の適用

以下に要する時間を計測する。

1. ファームウェアイメージ全体の生成（`time(1)`で計測する）
2. ファームウェアイメージのデバイスへの送信（`time(1)`で計測する）
3. デバイスが再起動し`ping(1)`で通信が疎通すること
4. IoTアプリケーションと`nc(1)`でTCP通信が疎通すること

3回計測を行い、それらの平均を用いる。

#### 1回目

In [5]:
!cd target_app && ../measure.sh firmware

nerves.local is at 172.31.68.221

Time to build firmware (mix firmware)

real	0m28.356s
user	0m6.683s
sys	0m2.310s

Time to upload firmware

real	0m14.359s
user	0m5.220s
sys	0m1.302s

Time to reboot

time: 22.728 sec.

Time to accept

Connection to 172.31.68.221 port 9849 [tcp/*] succeeded!
timeout count: 1
time: 1.056 sec.



ビルドされたファームウェアのサイズ:

In [7]:
!ls -lh target_app/_build/rpi3_dev/nerves/images/target_app.fw

-rw-r--r--  1 antipop  2033490572    42M  1 20 22:23 target_app/_build/rpi3_dev/nerves/images/target_app.fw


#### 2回目

In [10]:
!cd target_app && ../measure.sh firmware

nerves.local is at 172.31.68.221

Time to build firmware (mix firmware)

real	0m25.451s
user	0m6.377s
sys	0m2.085s

Time to upload firmware

real	0m13.922s
user	0m5.207s
sys	0m1.251s

Time to reboot

time: 22.610 sec.

Time to accept

Connection to 172.31.68.221 port 9849 [tcp/*] succeeded!
timeout count: 1
time: 1.053 sec.



ビルドされたファームウェアのサイズ:

In [11]:
!ls -lh target_app/_build/rpi3_dev/nerves/images/target_app.fw

-rw-r--r--  1 antipop  2033490572    42M  1 20 22:28 target_app/_build/rpi3_dev/nerves/images/target_app.fw


#### 3回目

In [14]:
!cd target_app && ../measure.sh firmware

nerves.local is at 172.31.68.221

Time to build firmware (mix firmware)

real	0m28.092s
user	0m6.256s
sys	0m2.211s

Time to upload firmware

real	0m15.227s
user	0m5.348s
sys	0m1.292s

Time to reboot

time: 22.767 sec.

Time to accept

Connection to 172.31.68.221 port 9849 [tcp/*] succeeded!
timeout count: 1
time: 1.042 sec.



ビルドされたファームウェアのサイズ:

In [15]:
!ls -lh target_app/_build/rpi3_dev/nerves/images/target_app.fw

-rw-r--r--  1 antipop  2033490572    42M  1 20 22:34 target_app/_build/rpi3_dev/nerves/images/target_app.fw


----

#### 方式（1）のまとめ

In [18]:
print("（1） {:.2f}秒".format((28.356+25.451+28.092)/3))
print("（2） {:.2f}秒".format((14.359+13.922+15.227)/3))
print("（3） {:.2f}秒".format((22.728+22.610+22.767)/3))
print("（4） {:.2f}秒".format((1.056+1.053+1.042)/3))

（1） 27.30秒
（2） 14.50秒
（3） 22.70秒
（4） 1.05秒


また、ビルドされたファームウェアのサイズは`42MB`であった。

----

### 2. ファームウェアイメージ差分の適用

以下に要する時間を計測する。

1. ファームウェアイメージ差分の生成（`time(1)`で計測する）
2. ファームウェアイメージのデバイスへの送信（`time(1)`で計測する）
3. デバイスが再起動し`ping(1)`で通信が疎通すること
4. IoTアプリケーションと`nc(1)`でTCP通信が疎通すること

3回計測を行い、それらの平均を用いる。

#### 1回目

In [8]:
!cd target_app && ../measure.sh firmware.patch

nerves.local is at 172.31.68.221

Time to build firmware (mix firmware.patch)

real	0m30.348s
user	0m6.424s
sys	0m2.342s

Time to upload firmware

real	0m17.616s
user	0m4.103s
sys	0m1.170s

Time to reboot

time: 22.645 sec.

Time to accept

Connection to 172.31.68.221 port 9849 [tcp/*] succeeded!
timeout count: 1
time: 1.055 sec.



ビルドされたファームウェア差分のサイズ:

In [9]:
!ls -lh target_app/_build/rpi3_dev/nerves/images/patch.fw

-rw-r--r--  1 antipop  2033490572   5.9M  1 20 22:26 target_app/_build/rpi3_dev/nerves/images/patch.fw


#### 2回目

In [12]:
!cd target_app && ../measure.sh firmware.patch

nerves.local is at 172.31.68.221

Time to build firmware (mix firmware.patch)

real	0m29.028s
user	0m6.118s
sys	0m2.183s

Time to upload firmware

real	0m17.130s
user	0m3.772s
sys	0m1.023s

Time to reboot

time: 22.582 sec.

Time to accept

Connection to 172.31.68.221 port 9849 [tcp/*] succeeded!
timeout count: 1
time: 1.045 sec.



ビルドされたファームウェア差分のサイズ:

In [13]:
!ls -lh target_app/_build/rpi3_dev/nerves/images/patch.fw

-rw-r--r--  1 antipop  2033490572   6.0M  1 20 22:31 target_app/_build/rpi3_dev/nerves/images/patch.fw


#### 3回目

In [17]:
!cd target_app && ../measure.sh firmware.patch

nerves.local is at 172.31.68.221

Time to build firmware (mix firmware.patch)

real	0m31.053s
user	0m6.576s
sys	0m2.411s

Time to upload firmware

real	0m17.459s
user	0m3.902s
sys	0m1.099s

Time to reboot

time: 22.662 sec.

Time to accept

Connection to 172.31.68.221 port 9849 [tcp/*] succeeded!
timeout count: 1
time: 1.052 sec.



ビルドされたファームウェア差分のサイズ:

In [22]:
!ls -lh target_app/_build/rpi3_dev/nerves/images/patch.fw

-rw-r--r--  1 antipop  2033490572   6.0M  1 20 22:42 target_app/_build/rpi3_dev/nerves/images/patch.fw


----

#### 方式（2）のまとめ

In [21]:
print("（1） {:.2f}秒".format((30.348+29.028+31.053)/3))
print("（2） {:.2f}秒".format((17.616+17.130+17.459)/3))
print("（3） {:.2f}秒".format((22.645+22.582+22.662)/3))
print("（4） {:.2f}秒".format((1.055+1.045+1.052)/3))

（1） 30.14秒
（2） 17.40秒
（3） 22.63秒
（4） 1.05秒


また、ビルドされたファームウェア差分のサイズは`6MB`であった。

----

### 3. アプリケーションコードの部分適用

この方式ではIoTデバイスの再起動は必要ないため、以下に要する時間を計測する。

1. コードの変更のデバイスへの適用（`time(1)`で計測する）
4. IoTアプリケーションと`nc(1)`でTCP通信が疎通すること

3回計測を行い、それらの平均を用いる。

#### 1回目

In [11]:
!cd target_app && time mix upload.hotswap > /dev/null


real	0m3.414s
user	0m4.708s
sys	0m1.322s


#### 2回め

In [15]:
!cd target_app && time mix upload.hotswap > /dev/null


real	0m3.197s
user	0m4.668s
sys	0m1.256s


#### 3回目

In [16]:
!cd target_app && time mix upload.hotswap > /dev/null


real	0m2.915s
user	0m4.528s
sys	0m1.200s


#### 方式（3）のまとめ

print("1. `mix upload.hotswap`: {:.2f}秒".format((3.414+3.197+2.915)/3))

### 各方式の計測結果のまとめ

各方式について、上記から以下の通り結果をまとめた。

|                         | firmware build | firmware upload | reboot | code update | total   |
|-------------------------|----------------|-----------------|--------|-------------|-------|
| firmware update (full)  |          21.19 |           14.48 |  22.70 |           - | 58.37 |
| firmware update (patch) |          24.38 |           18.01 |  22.64 |           - | 65.04 |
| proposed method         |              - |               - |      - |        3.18 |  3.18 |

----

以降は予備的な実験および考察である．

## 方式（3）についての追加実験

上記の方式（3）は，IoTデバイスの再起動をともなわない方式である．一方で，Erlang VM上でのコードの更新は行われるため，以下の2つの面で通信の遮断が発生するかどうかを確認する．

1. 適用前後のデバイスとの通信を`ping(1)`で検証する
2. 適用前後のIoTアプリケーションとの通信を検証する

### 1. 適用前後のデバイスとの通信を`ping(1)`で検証する

上記の適用を行う前後で、通信の遮断が発生するかどうかを、`ping(1)`コマンドにより検証する。

ターミナルを2つ開いておき、一方では以下のコマンドを用いて、`nerves.local`に対して0.1秒間隔で`ping(1)`コマンドを実行する（1）。

```
ping -i 0.1 nerves.local | while read pi; do echo "$(date '+[%Y/%m/%d %H:%M:%S]') $pi"; done
```

他方では、上述の`mix upload.hotswap`コマンドを`date(1)`コマンドを前後に挟むことで時間を表示して実行する（2）。

（1）の結果は以下の通りである。

```
$ ping -i 0.1 nerves.local | while read pi; do echo "$(date '+[%Y/%m/%d %H:%M:%S]') $pi"; done
[2021/01/15 16:54:24] PING nerves.local (172.31.68.221): 56 data bytes
[2021/01/15 16:54:24] 64 bytes from 172.31.68.221: icmp_seq=0 ttl=64 time=0.726 ms
[2021/01/15 16:54:24] 64 bytes from 172.31.68.221: icmp_seq=1 ttl=64 time=0.651 ms
[2021/01/15 16:54:24] 64 bytes from 172.31.68.221: icmp_seq=2 ttl=64 time=0.833 ms
[2021/01/15 16:54:24] 64 bytes from 172.31.68.221: icmp_seq=3 ttl=64 time=0.920 ms
[2021/01/15 16:54:25] 64 bytes from 172.31.68.221: icmp_seq=4 ttl=64 time=0.977 ms
[2021/01/15 16:54:25] 64 bytes from 172.31.68.221: icmp_seq=5 ttl=64 time=0.668 ms
[2021/01/15 16:54:25] 64 bytes from 172.31.68.221: icmp_seq=6 ttl=64 time=0.840 ms
[2021/01/15 16:54:25] 64 bytes from 172.31.68.221: icmp_seq=7 ttl=64 time=0.832 ms
[2021/01/15 16:54:25] 64 bytes from 172.31.68.221: icmp_seq=8 ttl=64 time=0.783 ms
[2021/01/15 16:54:25] 64 bytes from 172.31.68.221: icmp_seq=9 ttl=64 time=0.774 ms
[2021/01/15 16:54:25] 64 bytes from 172.31.68.221: icmp_seq=10 ttl=64 time=0.858 ms
[2021/01/15 16:54:25] 64 bytes from 172.31.68.221: icmp_seq=11 ttl=64 time=0.821 ms
[2021/01/15 16:54:25] 64 bytes from 172.31.68.221: icmp_seq=12 ttl=64 time=0.865 ms
[2021/01/15 16:54:26] 64 bytes from 172.31.68.221: icmp_seq=13 ttl=64 time=0.731 ms
[2021/01/15 16:54:26] 64 bytes from 172.31.68.221: icmp_seq=14 ttl=64 time=0.770 ms
[2021/01/15 16:54:26] 64 bytes from 172.31.68.221: icmp_seq=15 ttl=64 time=0.624 ms
[2021/01/15 16:54:26] 64 bytes from 172.31.68.221: icmp_seq=16 ttl=64 time=0.719 ms
[2021/01/15 16:54:26] 64 bytes from 172.31.68.221: icmp_seq=17 ttl=64 time=0.810 ms
[2021/01/15 16:54:26] 64 bytes from 172.31.68.221: icmp_seq=18 ttl=64 time=0.685 ms
[2021/01/15 16:54:26] 64 bytes from 172.31.68.221: icmp_seq=19 ttl=64 time=0.781 ms
[2021/01/15 16:54:26] 64 bytes from 172.31.68.221: icmp_seq=20 ttl=64 time=0.742 ms
[2021/01/15 16:54:26] 64 bytes from 172.31.68.221: icmp_seq=21 ttl=64 time=0.816 ms
[2021/01/15 16:54:26] 64 bytes from 172.31.68.221: icmp_seq=22 ttl=64 time=0.583 ms
[2021/01/15 16:54:27] 64 bytes from 172.31.68.221: icmp_seq=23 ttl=64 time=0.547 ms
[2021/01/15 16:54:27] 64 bytes from 172.31.68.221: icmp_seq=24 ttl=64 time=0.618 ms
[2021/01/15 16:54:27] 64 bytes from 172.31.68.221: icmp_seq=25 ttl=64 time=0.592 ms
[2021/01/15 16:54:27] 64 bytes from 172.31.68.221: icmp_seq=26 ttl=64 time=0.556 ms
[2021/01/15 16:54:27] 64 bytes from 172.31.68.221: icmp_seq=27 ttl=64 time=0.525 ms
[2021/01/15 16:54:27] 64 bytes from 172.31.68.221: icmp_seq=28 ttl=64 time=0.498 ms
[2021/01/15 16:54:27] 64 bytes from 172.31.68.221: icmp_seq=29 ttl=64 time=0.506 ms
[2021/01/15 16:54:27] 64 bytes from 172.31.68.221: icmp_seq=30 ttl=64 time=0.504 ms
[2021/01/15 16:54:27] 64 bytes from 172.31.68.221: icmp_seq=31 ttl=64 time=0.514 ms
[2021/01/15 16:54:27] 64 bytes from 172.31.68.221: icmp_seq=32 ttl=64 time=0.558 ms
[2021/01/15 16:54:28] 64 bytes from 172.31.68.221: icmp_seq=33 ttl=64 time=0.556 ms
[2021/01/15 16:54:28] 64 bytes from 172.31.68.221: icmp_seq=34 ttl=64 time=0.525 ms
[2021/01/15 16:54:28] 64 bytes from 172.31.68.221: icmp_seq=35 ttl=64 time=0.559 ms
[2021/01/15 16:54:28] 64 bytes from 172.31.68.221: icmp_seq=36 ttl=64 time=0.547 ms
[2021/01/15 16:54:28] 64 bytes from 172.31.68.221: icmp_seq=37 ttl=64 time=0.517 ms
[2021/01/15 16:54:28] 64 bytes from 172.31.68.221: icmp_seq=38 ttl=64 time=0.518 ms
[2021/01/15 16:54:28] 64 bytes from 172.31.68.221: icmp_seq=39 ttl=64 time=0.534 ms
[2021/01/15 16:54:28] 64 bytes from 172.31.68.221: icmp_seq=40 ttl=64 time=0.617 ms
[2021/01/15 16:54:28] 64 bytes from 172.31.68.221: icmp_seq=41 ttl=64 time=0.530 ms
[2021/01/15 16:54:28] 64 bytes from 172.31.68.221: icmp_seq=42 ttl=64 time=0.581 ms
[2021/01/15 16:54:29] 64 bytes from 172.31.68.221: icmp_seq=43 ttl=64 time=0.633 ms
[2021/01/15 16:54:29] 64 bytes from 172.31.68.221: icmp_seq=44 ttl=64 time=0.534 ms
[2021/01/15 16:54:29] 64 bytes from 172.31.68.221: icmp_seq=45 ttl=64 time=0.533 ms
[2021/01/15 16:54:29] 64 bytes from 172.31.68.221: icmp_seq=46 ttl=64 time=0.548 ms
[2021/01/15 16:54:29] 64 bytes from 172.31.68.221: icmp_seq=47 ttl=64 time=0.520 ms
[2021/01/15 16:54:29] 64 bytes from 172.31.68.221: icmp_seq=48 ttl=64 time=0.561 ms
[2021/01/15 16:54:29] 64 bytes from 172.31.68.221: icmp_seq=49 ttl=64 time=0.490 ms
[2021/01/15 16:54:29] 64 bytes from 172.31.68.221: icmp_seq=50 ttl=64 time=0.524 ms
[2021/01/15 16:54:29] 64 bytes from 172.31.68.221: icmp_seq=51 ttl=64 time=0.501 ms
[2021/01/15 16:54:29] 64 bytes from 172.31.68.221: icmp_seq=52 ttl=64 time=0.639 ms
[2021/01/15 16:54:30] 64 bytes from 172.31.68.221: icmp_seq=53 ttl=64 time=0.735 ms
[2021/01/15 16:54:30] 64 bytes from 172.31.68.221: icmp_seq=54 ttl=64 time=0.771 ms
[2021/01/15 16:54:30] 64 bytes from 172.31.68.221: icmp_seq=55 ttl=64 time=0.815 ms
[2021/01/15 16:54:30] 64 bytes from 172.31.68.221: icmp_seq=56 ttl=64 time=0.841 ms
[2021/01/15 16:54:30] 64 bytes from 172.31.68.221: icmp_seq=57 ttl=64 time=0.777 ms
[2021/01/15 16:54:30] 64 bytes from 172.31.68.221: icmp_seq=58 ttl=64 time=0.701 ms
[2021/01/15 16:54:30] 64 bytes from 172.31.68.221: icmp_seq=59 ttl=64 time=0.688 ms
[2021/01/15 16:54:30] 64 bytes from 172.31.68.221: icmp_seq=60 ttl=64 time=0.596 ms
[2021/01/15 16:54:30] 64 bytes from 172.31.68.221: icmp_seq=61 ttl=64 time=0.755 ms
[2021/01/15 16:54:31] 64 bytes from 172.31.68.221: icmp_seq=62 ttl=64 time=0.586 ms
[2021/01/15 16:54:31] 64 bytes from 172.31.68.221: icmp_seq=63 ttl=64 time=0.543 ms
[2021/01/15 16:54:31] 64 bytes from 172.31.68.221: icmp_seq=64 ttl=64 time=0.552 ms
[2021/01/15 16:54:31] 64 bytes from 172.31.68.221: icmp_seq=65 ttl=64 time=0.760 ms
[2021/01/15 16:54:31] 64 bytes from 172.31.68.221: icmp_seq=66 ttl=64 time=0.844 ms
[2021/01/15 16:54:31] 64 bytes from 172.31.68.221: icmp_seq=67 ttl=64 time=0.870 ms
^C
```

（2）の結果は以下の通りである。

```
$ date '+[%Y/%m/%d %H:%M:%S]'; mix upload.hotswap; date '+[%Y/%m/%d %H:%M:%S]'
[2021/01/15 16:54:26]
==> nerves
make[1]: Nothing to be done for `all'.
==> target_app

Nerves environment
  MIX_TARGET:   rpi3
  MIX_ENV:      dev

Compiling 2 files (.ex)
Generated target_app app
Successfully connected to hot_upload_test@nerves.local
Successfully deployed Elixir.TargetApp to hot_upload_test@nerves.local
Successfully deployed Elixir.TargetApp.Application to hot_upload_test@nerves.local
[2021/01/15 16:54:29]
```

よって、（1）の`ping(1)`の実行中に、（2）のコマンド実行による通信の切断は見られなかった。